### PR TITLE
En 5778/hard fork tries

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: integrationTests/state/stateTrieSync_test.go [hunk 1]
```diff
@@ -51,10 +55,21 @@ func TestNode_RequestInterceptTrieNodesWithMessenger(t *testing.T) {
...
+	}})
+	requestHandler, _ := requestHandlers.NewResolverRequestHandler(
+		nRequester.ResolverFinder,
+		&mock.RequestedItemsHandlerStub{},
+		whiteListHandler,
+		10000,
+		nRequester.ShardCoordinator.SelfId(),
+	)
+
+	_ = logger.SetLogLevel("*:TRACE")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/state/stateTrieSync_test.go]: done

File: process/block/shardblock.go [hunk 2]
```diff
@@ -921,6 +923,38 @@ func (sp *shardProcessor) updateState(headers []data.HeaderHandler) {
...
+	accounts := sp.accountsDB[state.UserAccountsState]
+	if !accounts.IsPruningEnabled() {
+		return
+	}
+
+	if header.IsStartOfEpochBlock() {
+		epochStartId := core.EpochStartIdentifier(header.GetEpoch())
+		metaBlock, err := process.GetMetaHeaderFromStorage([]byte(epochStartId), sp.marshalizer, sp.store)
+		if err != nil {
+			log.Warn("could not find epoch start metablock for", "epoch", header.GetEpoch(), "err", err)
```
Comment (iulianpascalau) [inline process/block/shardblock.go]: log.Debug? No validator cares about this warn.

File: process/block/shardblock.go [hunk 3]
```diff
@@ -921,6 +923,38 @@ func (sp *shardProcessor) updateState(headers []data.HeaderHandler) {
...
+
+			rootHash := epochStartShData.RootHash
+			accounts.CancelPrune(rootHash, data.NewRoot)
+			log.Debug("shard trie snapshot from epoch start shard data", "rootHash", rootHash)
+			accounts.SnapshotState(rootHash)
+
+			return
+		}
+
+		log.Warn("could not find epoch start shard data in metaBlock for", "epoch", header.GetEpoch(), "err", err)
```
Comment (iulianpascalau) [inline process/block/shardblock.go]: log.Debug

File: update/factory/trieSyncersContainerFactory.go [hunk 4]
```diff
@@ -121,6 +109,16 @@ func (t *trieSyncersContainerFactory) createOneTrieSyncer(
 	return nil
 }
 
+func trieTopicFromAccountType(accType factory.Type) string {
```
Comment (iulianpascalau) [inline update/factory/trieSyncersContainerFactory.go]: suggestion to move this function + factory.Type + factoryTrie.topics in core

File: update/sync/syncTries.go [hunk 5]
```diff
@@ -107,12 +108,12 @@ func (st *syncTries) SyncTriesFrom(meta *block.MetaBlock, waitTime time.Duration
...
 
 func (st *syncTries) syncMeta(meta *block.MetaBlock) error {
-	err := st.syncTrieOfType(factory.UserAccount, core.MetachainShardId, meta.RootHash)
+	err := st.syncTrieOfType(factory.UserAccount, factoryTrieContainer.UserAccountTrie, core.MetachainShardId, meta.RootHash)
 	if err != nil {
 		return err
 	}
 
-	err = st.syncTrieOfType(factory.ValidatorAccount, core.MetachainShardId, meta.ValidatorStatsRootHash)
+	err = st.syncTrieOfType(factory.ValidatorAccount, factoryTrieContainer.PeerAccountTrie, core.MetachainShardId, meta.ValidatorStatsRootHash)
```
Comment (iulianpascalau) [inline update/sync/syncTries.go]: Peer <> Valiator? Suggestion to have only one of them

File: process/block/shardblock.go (hunk 2 above)
Comment (sasurobert) [reply process/block/shardblock.go]: I would leave as log.Warn. it should never happen, only if there was something incorrectly coded.

File: process/block/shardblock.go (hunk 3 above)
Comment (sasurobert) [reply process/block/shardblock.go]: I would leave as log.Warn. it should never happen, only if there was something incorrectly coded.

File: update/factory/trieSyncersContainerFactory.go (hunk 4 above)
Comment (sasurobert) [reply update/factory/trieSyncersContainerFactory.go]: only used in hardfork.

File: update/sync/syncTries.go (hunk 5 above)
Comment (sasurobert) [reply update/sync/syncTries.go]: agreed. but would do it in development to not have too many merge conflicts.

File: process/block/shardblock.go (hunk 2 above)
Comment (iulianpascalau) [reply process/block/shardblock.go]: ok

File: process/block/shardblock.go (hunk 3 above)
Comment (iulianpascalau) [reply process/block/shardblock.go]: ok

**Pull request merged**